enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,879 @@
+other resources created will also be cleaned up.
+
+A fork of [upstream](https://github.com/cert-manager/istio-csr) `istio-csr` will be created [downstream](https://github.com/openshift)
+for better version management.
+
+### Workflow Description
+
+- An OpenShift user enables `istio-csr` by creating the new custom resource made available for configuring
+  `istio-csr`.
+- `istio-csr` controller based on the configuration in CR
+  - Creates below `Issuers` when user does not provide `Issuer Reference` to be used by `istio-csr` for
+    obtaining the certificates.
+    - Creates an `ClusterIssuer` in `cert-manager` namespace which acts as root CA for obtaining the
+      CA certificate for `istio-csr` use case.
+    - Creates an `Issuer` in `istio-system` namespace which is an intermediate CA signed by `ClusterIssuer`
+      root CA.
+  - Deploys istio-csr agent in `istio-system` namespace.
+
+### API Extensions
+
+Below new API `istiocsrs.operator.openshift.io` is introduced for managing istio-csr.
+```golang
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// IstioCSR describes a managed istio-csr agent for the cluster. The
+// controller can service requests to obtain certificates from the configured
+// issuer.
+//
+// When an IstioCSR is created, a new deployment is created which manages the
+// istio-csr agent and keeps it in the desired state.
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'default'",message="istiocsr is a singleton, .metadata.name must be 'default'"
+type IstioCSR struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the IstioCSR.
+	Spec IstioCSRSpec `json:"spec,omitempty"`
+
+	// status is the most recently observed status of the IstioCSR.
+	Status IstioCSRStatus `json:"status,omitempty"`
+}
+
+// IstioCSRSpec is the specification of the desired behavior of the IstioCSR.
+type IstioCSRSpec struct {
+	apiv1.OperatorSpec `json:",inline"`
+
+	// ControllerConfig is for configuring istio-csr operator behavior.
+	ControllerConfig `json:"controllerConfig,omitempty"`
+
+	// ApplicationConfig is for configuring istio-csr operand behavior.
+	ApplicationConfig `json:"applicationConfig,omitempty"`
+}
+
+// ControllerConfig is for configuring istio-csr operator behavior.
+type ControllerConfig struct {
+	// Namespace where the istio control-plane is running.
+	// Same namespace will be used for deploying istio-csr agent.
+	// +kubebuilder:default:="istio-system"
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="namespace is immutable once set"
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
```

Comment:
But `Namespace` is for installing the operand `istio-csr`, the controller is managing. And thought not assume the `istiod` namespace, made it as a config param. Please let me know WDYT.

Reply:
I am not sure that I understand it correctly. If I create `IstioCSR` resource, will it result in deploying a controller that will manage `istio-csr` agent? And the controller will be created in `IstioCSR.metadata.namespace` and the agent in `IstioCSR.spec.controllerConfig.namespace`?

Reply:
Yeah the field name `ControllerConfig` is misleading, I will update it.

> the controller will be created in IstioCSR.metadata.namespace and the agent in IstioCSR.spec.controllerConfig.namespace

istio-csr manager is part of `cert-manager-operator` which will watch for `istiocsr` custom resource and install `istio-csr` agent in `IstioCSR.spec.controllerConfig.namespace`. In fact all the configs in `controllerConfig` is for setting up the agent. I will update the name and comments appropriately.
And I think it would be better to add a CEL validation for restricting the `istiocsr` custom resource in cert-manager-operator namespace. Please let me know WDYT.

Reply:
If istio-csr agent must run in the same namespace as istiod, I would remove `controllerConfig.namespace` and rely only on the `metadata.namespace`.

Reply:
Yeah, but the expectation is to create the new CR in operator namespace and namespace where istiod will be installed need not exist at the time, operator will create it.

Reply:
> namespace where istiod will be installed need not exist at the time

Why?

Reply:
My understanding is based on this [doc](https://cert-manager.io/docs/usage/istio-csr/#installing-istio-csr-after-istio) that, istio-csr can be and must be installed before istio. Please correct me if understanding is wrong.